lib/portlayer/storage/vsphere/image.go

```diff
@@ -283,6 +246,124 @@ func (v *ImageStore) WriteImage(ctx context.Context, parent *portlayer.Image, ID
 	return newImage, nil
 }
 
+// Create a temporary directory, create a vmdk in this directory, attach/mount
+// the disk, unpack the tar, check the checksum.  If the data doesn't match the
+// expected checksum, abort by nuking the tempdir.  If everything matches, move
+// the tmpdir to the expected location (with the vmdk inside it).  The unwind
+// path is a bit convoluted here;  we need to clean up on the way out in the
+// error case (using the tmpdir).
+func (v *ImageStore) writeImage(ctx context.Context, storeName, parentID, ID,
+	sum string, r io.Reader) error {
+
+	// Create a temp image directory in the store.
+	tmpImageDir := v.imageDirPath(storeName, ID+"inprogress")
+	_, err := v.ds.Mkdir(ctx, true, tmpImageDir)
+	if err != nil {
+		return err
+	}
+
+	// datastore path to the parent
+	parentDiskDsURI := v.imageDiskPath(storeName, parentID)
+
+	// datastore path to the disk we're creating
+	imageDiskDsURI := path.Join(v.ds.RootURL, tmpImageDir, ID+".vmdk")
+	log.Infof("Creating image %s (%s)", ID, imageDiskDsURI)
+
+	// Create the disk
+	vmdisk, err := v.dm.CreateAndAttach(ctx, imageDiskDsURI, parentDiskDsURI, 0, os.O_RDWR)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if vmdisk.Mounted() {
+			log.Debugf("Unmounting abandonned disk")
+			vmdisk.Unmount()
+		}
+		if vmdisk.Attached() {
+			log.Debugf("Detaching abandonned disk")
+			v.dm.Detach(ctx, vmdisk)
+		}
+
+		v.ds.Rm(ctx, tmpImageDir)
+	}()
+
+	// tmp dir to mount the disk
+	dir, err := ioutil.TempDir("", "mnt-"+ID)
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(dir)
+
+	if err := vmdisk.Mount(dir, nil); err != nil {
+		return err
+	}
+
+	h := sha256.New()
+	t := io.TeeReader(r, h)
+
+	// Untar the archive
+	if err = archive.Untar(t, dir, &archive.TarOptions{}); err != nil {
+		return err
+	}
+
+	actualSum := fmt.Sprintf("sha256:%x", h.Sum(nil))
+	if actualSum != sum {
+		return fmt.Errorf("Failed to validate image checksum. Expected %s, got %s", sum, actualSum)
+	}
+
+	if err = vmdisk.Unmount(); err != nil {
+		return err
+	}
+
+	if err = v.dm.Detach(ctx, vmdisk); err != nil {
+		return err
+	}
+
+	// If we've gotten here, prepare the final location for the image
+	imageDir := v.imageDirPath(storeName, ID)
+	_, err = v.ds.Mkdir(ctx, true, imageDir)
+	if err != nil {
```

Comment:
what happens if we previously died at this point, having created the directory but not moved the image in?